### PR TITLE
Introduce configurable OTP properties with flow-level overrides

### LINKED ISSUE #3425: Make OTP parameters configurable per flow (flow-level override)
### Description
  Building on #3297 (server-level configuration, merged), OTP generation parameters — length, character set, validity period — should also be configurable per flow via executor node properties. This is the next layer in the agreed override hierarchy: **server →  flow**.
  
  ### Context
  The notification-provider-level layer originally proposed in #3356 was skipped per design alignment: notification providers are global, and per-scenario security policy is better expressed at the flow level. The resolver pattern from Phase 1 accommodates flow-level overrides directly.
  
  ### Related
  - Server-level config: #3297 (merged)
  - Provider-level issue : #3356
  - Design discussion: #3233

### PR DESCRIPTION
**Pull request by HesandaLiyanage**
### Purpose

Add flow-level OTP configuration overrides (length, character set, validity period) to the OTP executor, allowing flow authors to override the global `notification.otp.*` defaults per flow node.

Design discussion: [#3233](https://github.com/thunder-id/thunderid/issues/3425) — Phase 2 redirected to flow-level, per alignment with @rajithacharith and @ThaminduDilshan. Skips the notification-provider layer and goes direct to flow-level node properties using the existing `propertyKey*` pattern.

### Approach

Three OTP property keys added to `flow/executor/constants.go`:
- `propertyKeyOTPLength` — OTP digit length (`[4, 10]`)
- `propertyKeyOTPUseNumericOnly` — numeric-only charset toggle
- `propertyKeyOTPValidityPeriodSeconds` — validity in seconds (`[30, 600]`)

**Resolution behaviour:**
- Node properties set → use them (full or partial override).
- Node properties unset → fall back to server config.
- Partial node config → use node values where set, server config for the rest.

**Plumbing:**
1. `OTPConfig` struct (pointer fields for partial overrides) added to `notification/common/model.go`, aliased as `notification.OTPConfig`.
2. `SendOTPDTO` gains an `OTPConfigOverride *OTPConfig` field.
3. `resolveOTPConfig()` takes an optional `*OTPConfig` and merges non-nil fields over the global config.
4. `otpExecutor.resolveOTPConfigOverride()` reads NodeProperties and constructs the override; passes it through `otpAuthnService.GenerateOTP()` → `notifOTPService.GenerateOTP()` → `generateOTP()` + accessor methods.
5. Console API path (`authenticationService.SendOTP`) passes `nil` — no overrides on that path.

### Related Issues
- https://github.com/thunder-id/thunderid/issues/3425

### Related PRs
- N/A

### Checklist
- [x] Followed the contribution guidelines.
- [ ] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
    - [ ] Ran Vale and fixed all errors and warnings
- [x] Tests provided. (Add links if there are any)
    - [x] Unit Tests
    - [ ] Integration Tests
- [ ] Breaking changes. (Fill if applicable)
    - [ ] Breaking changes section filled.
    - [ ] `breaking change` label added.

### Security checks
- [x] Followed secure coding standards in [WSO2 Secure Coding Guidelines](https://security.docs.wso2.com/en/latest/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/introduction/)
- [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.

## Summary by CodeRabbit

* **New Features**
  * Added configurable OTP settings for flow steps, including code length, numeric-only mode, and validity period.
  * OTP configuration overrides now apply consistently during code generation and delivery.
  * Added support for converting common boolean values from configuration inputs.

* **Bug Fixes**
  * Invalid OTP configuration values are safely rejected while preserving existing fallback behavior.

* **Tests**
  * Added coverage for valid, invalid, partial, and boundary OTP configuration scenarios.

### CODE REVIEW

File: backend/internal/authn/service.go [hunk 1]
```diff
@@ -209,7 +209,7 @@ func (as *authenticationService) SendOTP(ctx context.Context, senderID string, c
 	recipient string) (string, *tidcommon.ServiceError) {
 	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, svcLoggerComponentName))
 
-	sessionToken, otpValue, _, svcErr := as.otpService.GenerateOTP(ctx, recipient, "mobile_number")
+	sessionToken, otpValue, _, svcErr := as.otpService.GenerateOTP(ctx, recipient, "mobile_number", nil)
```
[the remark this answers is not in the document]
Comment (HesandaLiyanage) [reply backend/internal/authn/service.go:211]: This is a pre existing design issue, not introduced by this change.

File: backend/internal/flow/executor/constants.go [hunk 2]
```diff
@@ -115,6 +115,9 @@ const (
 	propertyKeyCallbackType                            = "callbackType"
 	propertyKeyLoginHintAttribute                      = "loginHintAttribute"
 	propertyKeyMaxOTPAttempts                          = "maxAttempts"
+	propertyKeyOTPLength                               = "otpLength"
+	propertyKeyOTPUseNumericOnly                       = "otpUseNumericOnly"
+	propertyKeyOTPValidityPeriodSeconds                = "otpValidityPeriodSeconds"
```
[the remark this answers is not in the document]
Comment (HesandaLiyanage) [reply backend/internal/flow/executor/constants.go:120]: Done

File: backend/internal/flow/executor/otp_executor.go [hunk 3]
```diff
@@ -370,6 +372,63 @@ func (e *otpExecutor) validateAttempts(ctx *providers.NodeContext, execResp *pro
...
+			hasOverride = true
+		}
+	}
+
+	if !hasOverride {
+		return nil
+	}
+	return &cfg
+}
+
```
[the remark this answers is not in the document]
Comment (HesandaLiyanage) [reply backend/internal/flow/executor/otp_executor.go]: Done

File: backend/internal/flow/executor/otp_executor.go [hunk 4]
```diff
@@ -378,6 +373,63 @@ func (e *otpExecutor) validateAttempts(ctx *providers.NodeContext, execResp *pro
...
+// toInt converts a NodeProperty value to int, supporting string, int, and float64 types.
+func toInt(v interface{}) (int, error) {
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/otp_executor.go]: We should be able to reuse `sysutils.ToInt64()` from backend/internal/system/utils/value_util.go. No need for a new method

File: backend/internal/flow/executor/otp_executor.go [hunk 5]
```diff
@@ -378,6 +373,63 @@ func (e *otpExecutor) validateAttempts(ctx *providers.NodeContext, execResp *pro
...
+	case int:
+		return val, nil
+	case float64:
+		return int(val), nil
+	}
+	return 0, fmt.Errorf("cannot convert %T to int", v)
+}
+
+// toBool converts a NodeProperty value to bool, supporting string and bool types.
+func toBool(v interface{}) (bool, error) {
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/otp_executor.go]: Shall we add a new method to backend/internal/system/utils/value_util.go for this rather than introducing the method here.
First check whether a method already exists in system/utils

File: backend/internal/notification/common/model.go [hunk 6]
```diff
@@ -84,11 +84,20 @@ type VerifyOTPResponse struct {
...
+
 // SendOTPDTO represents the service layer data structure for sending an OTP.
 type SendOTPDTO struct {
-	Recipient string
-	SenderID  string
-	Channel   string
+	Recipient         string
+	SenderID          string
+	Channel           string
+	OTPConfigOverride *OTPConfig
```
Comment (ThaminduDilshan) [inline backend/internal/notification/common/model.go]:
```suggestion
	OTPConfig *OTPConfig
```

File: backend/internal/notification/config.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package notification
+
+import "github.com/thunder-id/thunderid/internal/notification/common"
+
+// OTPConfig is an alias for the common package definition so it can be referenced
+// as notification.OTPConfig externally.
+type OTPConfig = common.OTPConfig
```
Comment (ThaminduDilshan) [inline backend/internal/notification/config.go]: No need for a alias. We can let it be imported directly from backend/internal/notification/common/model.go

File: backend/internal/notification/otp_service.go [hunk 8]
```diff
@@ -236,8 +238,9 @@ func (s *otpService) validateOTPVerifyRequest(request common.VerifyOTPDTO) *tidc
 
 // sendSMSOTP sends an SMS OTP to the recipient.
 func (s *otpService) sendSMSOTP(ctx context.Context, recipient, otpVal string,
-	sender common.NotificationSenderDTO, logger *log.Logger) *tidcommon.ServiceError {
-	otpCfg := s.resolveOTPConfig()
+	sender common.NotificationSenderDTO, logger *log.Logger,
+	otpOverride *common.OTPConfig) *tidcommon.ServiceError {
```
Comment (ThaminduDilshan) [inline backend/internal/notification/otp_service.go]:
```suggestion
	otpCfg *common.OTPConfig) *tidcommon.ServiceError {
```

File: backend/internal/notification/otp_service.go [hunk 9]
```diff
@@ -236,8 +238,9 @@ func (s *otpService) validateOTPVerifyRequest(request common.VerifyOTPDTO) *tidc
...
+	otpCfg := s.resolveOTPConfig(otpOverride)
```
Comment (ThaminduDilshan) [inline backend/internal/notification/otp_service.go]: We can use `getOTPValidityPeriodInMillis()` here right? Shall we update code that way?

**Comment by ThaminduDilshan** [comment]
@HesandaLiyanage Thanks for actively working on this task. I have added couple of review comments. Apart from that, can we also do the following;
1. Add tests for the otp config override and nil cases
2. These executor properties can be configured from the console flow builder UI. Can you work on adding new properties also to the UI? Require adding i18n keys as well

Also once done, can you add a screenshot/ recording confirming the fix? I'd like to see configuring these from flow builder UI and observing that in runtime (test OTP with different configs)

File: backend/internal/notification/common/model.go (hunk 6 above)
Comment (HesandaLiyanage) [reply backend/internal/notification/common/model.go]: Resolved and force pushed

File: backend/internal/notification/config.go (hunk 7 above)
Comment (HesandaLiyanage) [reply backend/internal/notification/config.go]: That file only contains that alias. Removed the file.

**Comment by rajithacharith** [comment]
@HesandaLiyanage seems there is a lint failure and patch coverage is missing, can you check this with merge conflicts as well.

File: .gitignore [hunk 10]
```diff
@@ -133,3 +133,9 @@ docs/static/api/next/postman/
 # Dev-only CORS seed staged by `build.sh run` / `build.ps1 run` for the bootstrap one-shot (never committed or packaged)
 backend/cmd/server/bootstrap/02-server-configurations.yaml
 
+# Nx build cache (generated, never committed)
+.nx/
+
+# Development internals guide (not intended for commit)
```
Comment (ThaminduDilshan) [inline .gitignore:142]: Do we need this?

**Comment by ThaminduDilshan** [comment]
@HesandaLiyanage are you still working on this issue?

**Comment by HesandaLiyanage** [comment]
@ThaminduDilshan 
Yes, still working on it. The UI changes for OTP config properties are done and the backend fixes are in. Currently testing the end-to-end flow to capture the recording you requested. Will update shortly

File: .gitignore (hunk 10 above)
Comment (HesandaLiyanage) [reply .gitignore:142]: .nx/ is a build cache, auto generated. It stores computed outputs for build/test/lint tasks. This entry is correct.

**Comment by HesandaLiyanage** [comment]
@rajithacharith @ThaminduDilshan 
Thanks for the review. The branch is behind upstream/main which includes a significant refactor of the OTP service (notification sender CRUD → /connections, removed resolveOTPConfigOverride, etc.). Our changes conflict with this. 

upstream/main removed the OTPConfig struct, resolveOTPConfigOverride(), and the otpCfg parameter from GenerateOTP — our current implementation conflicts with all of those. The upstream also introduced ExecutorMeta.SupportedProperties and a 5th *ExecutorMeta param on CreateExecutor, which we'll use.

The re-implementation follows the pattern already established for maxAttempts/getMaxOTPAttempts: resolve property overrides in the executor, fall back to global config, pass concrete values to the service. No pointer-struct merging in the service layer.

Specifically:
- otp_executor.go: new resolveOTPProperties() reads otpLength, otpUseNumericOnly, otpValidityPeriodSeconds from NodeProperties — all 3 registered in SupportedProperties
- otp_service.go + authn/otp/service.go: GenerateOTP accepts (length int, useNumericOnly bool, validitySeconds int) as params — no OTPConfig struct, no service-side merge, no resolveOTPConfig() override path
- Frontend UI (executors.json, OtpProperties.tsx, en-US.ts) applies clean — no changes needed

If this approach sounds right, I'll push the re-worked PR. If not we can discuss further more in the design discussion #3233

**Comment by ThaminduDilshan** [comment]
> otp_service.go + authn/otp/service.go: GenerateOTP accepts (length int, useNumericOnly bool, validitySeconds int) as params — no OTPConfig struct, no service-side merge, no resolveOTPConfig() override path

I think better to use a OTPConfig struct rather than adding multiple method arguments

**Comment by HesandaLiyanage** [comment]
> > otp_service.go + authn/otp/service.go: GenerateOTP accepts (length int, useNumericOnly bool, validitySeconds int) as params — no OTPConfig struct, no service-side merge, no resolveOTPConfig() override path
> 
> I think better to use a OTPConfig struct rather than adding multiple method arguments

Good call on the struct. 

Updated approach, a concrete OTPConfig struct (no nil-pointer merge pattern):
```
// In otp_service.go (or internal package, not common/model.go since upstream gutted that)
type OTPConfig struct {
    Length                int
    UseNumericOnly        bool
    ValidityPeriodSeconds int
}
```

Executor resolves all 3 fields against global defaults and returns a filled struct:
resolveOTPProperties(ctx) → OTPConfig  // no nil fields, no merge logic

Service receives it directly with no fallback logic:
GenerateOTP(ctx, recipient, recipientAttr, cfg OTPConfig)
generateOTP(cfg OTPConfig) → getOTPCharset(cfg.UseNumericOnly), getOTPLength(cfg.Length), ...

This keeps the same "resolve in executor, pass concrete values" architecture the upstream uses for maxAttempts, while making the signature tidy. resolveOTPConfig() in the service can be removed since it's no longer needed.

Good to go? I'll push the re-worked PR with this approach.

**Comment by ThaminduDilshan** [comment]
> Executor resolves all 3 fields against global defaults and returns a filled struct:
> Service receives it directly with no fallback logic:

Let's resolve node properties only in executor level. notificationSenderService can apply defaults when not overridden (either from notif sender props or server defaults). Reason is this same notificationSenderService can be invoked via other paths such as atomic auth APIs.

Other than that approach looks good. You can proceed with the implementation

File: backend/internal/notification/otp_config.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package notification
+
+// OTPConfig holds optional OTP generation overrides.
+// Nil pointer = use server default.
+type OTPConfig struct {
```
Comment (ThaminduDilshan) [inline backend/internal/notification/otp_config.go]: Shall we move this to notification/common/model.go rather than introducing a new file?

File: backend/internal/notification/otp_service.go [hunk 12]
```diff
@@ -233,25 +236,37 @@ func (s *otpService) verifyAndDecodeSessionToken(ctx context.Context, token stri
...
-}
-
 // getOTPValidityPeriodInMillis returns the OTP validity period in milliseconds.
-func (s *otpService) getOTPValidityPeriodInMillis() int64 {
-	return int64(s.resolveOTPConfig().ValidityPeriodSeconds) * 1000
-}
+func (s *otpService) getOTPValidityPeriodInMillis(otpCfg *OTPConfig) int64 {
+	mergedCfg := s.resolveOTPConfig(otpCfg)
+	return int64(mergedCfg.ValidityPeriodSeconds) * 1000
+}
```
Comment (ThaminduDilshan) [inline backend/internal/notification/otp_service.go]: An empty line should be added at end of each file

File: backend/internal/flow/executor/otp_executor.go [hunk 13]
```diff
@@ -378,6 +384,50 @@ func (e *otpExecutor) validateAttempts(ctx *providers.NodeContext, execResp *pro
...
+// the fields that were explicitly set and valid.
+func (e *otpExecutor) resolveOTPProperties(ctx *providers.NodeContext) *notification.OTPConfig {
+	var cfg notification.OTPConfig
+	hasOverride := false
+
+	if v, ok := ctx.NodeProperties[propertyKeyOTPLength]; ok {
+		if n, ok := systemutils.ToInt64(v); ok {
+			if f, isFloat := v.(float64); !isFloat || f == float64(n) {
+				length := int(n)
+				if length >= 4 && length <= 10 {
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/otp_executor.go]: Better to do this runtime validations at notification/otp_service. Applicable to propertyKeyOTPValidityPeriodSeconds as well

File: backend/internal/flow/executor/otp_executor.go [hunk 14]
```diff
@@ -378,6 +384,50 @@ func (e *otpExecutor) validateAttempts(ctx *providers.NodeContext, execResp *pro
...
+					cfg.Length = &length
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/otp_executor.go]: Also I agree with your directive to have validations for the node properties. But performing it at runtime silently ignores the invalid values without giving any feedback to the dev. IMO `flow/mgt` flowValidator is the right place for this. Ref: https://github.com/thunder-id/thunderid/blob/59ba52f89c4df8a5ad095759a61da55329d78f16/backend/internal/flow/mgt/validator.go#L924.

But with this, we should allow executor to define their own conditions. I think we can enrich `ExecutorSupportedProperties` struct to have a validation rule.

```go
type PropertyValidation struct {
    Type          PropertyType // "int" | "bool" | "string" | "enum"
    Min           *int64
    Max           *int64
    AllowedValues []string
}

type ExecutorSupportedProperties struct {
    Property        string
    IsRequired      bool
    ApplicableModes []string
    Validation      *PropertyValidation
}
```

We can take this to a followup task. @coderabbitai can you create an issue to track this?

Comment (HesandaLiyanage) [reply backend/internal/flow/executor/otp_executor.go]: @ThaminduDilshan 
Agreed , design time validation in validateExecutorProperties with declarative rules on ExecutorSupportedProperties is the cleaner fix. I've already addressed the other review comments (trailing newline, moved range validation to service). The service level validation in resolveOTPConfig stays as runtime defense for direct API callers. Should I pick up the followup issue as well?

Comment (ThaminduDilshan) [reply backend/internal/flow/executor/otp_executor.go]: It's available. If you want to implement that, I can assign it to you

**Comment by ThaminduDilshan** [comment]
Shall we resolve the conflict?

**Review by ThaminduDilshan — DISMISSED** [review]

**Comment by ThaminduDilshan** [comment]
Looks like there's a lint failure:
https://github.com/thunder-id/thunderid/actions/runs/30214674280/job/89826643443

**Comment by HesandaLiyanage** [comment]
> Looks like there's a lint failure: https://github.com/thunder-id/thunderid/actions/runs/30214674280/job/89826643443

On it

**Comment by HesandaLiyanage** [comment]
> Looks like there's a lint failure: https://github.com/thunder-id/thunderid/actions/runs/30214674280/job/89826643443

Fixed

**Review by ThaminduDilshan — APPROVED** [review]
Thanks for the contribution. We really appreciate your effort to make thunderid better by actively working on PRs ❤️

File: backend/internal/flow/executor/otp_executor.go (hunk 14 above)
Comment (HesandaLiyanage) [reply backend/internal/flow/executor/otp_executor.go]: > It's available. If you want to implement that, I can assign it to you

Thanks @ThaminduDilshan . Going to take a break from open source for now cus uni exams coming up. This was my first open source contribution and I learned a ton from the whole process. Really appreciate your guidance and patience with the reviews. Really enjoyed working on IAM. I'll hop back into ThunderID in a few weeks.
